Support JSON RPC API

ggicci opened the pull request:
Support JSON RPC API

ggicci committed:
Add JSON-RPC API serving through websocket

ggicci committed:
Add jsonrpc2.ObjectStream implementation over HTTP

ggicci committed:
Update jsonrpc2 and websocket dependencies

ggicci committed:
Suppress API service testing now

ggicci committed:
Add block and transaction models for the API

ggicci committed:
Remove method echo

ggicci committed:
Add middleware to pre-process RPC method parameters

ggicci committed:
Add rpc methods for blocks and transactions

ggicci committed:
Start json rpc websocket service with cqld bootstrap

ggicci committed:
Update dependencies, add gorp db helper

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

xq262144 committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Post processing data retrieved from db

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Don't start JSON-RPC server if wsapiAddr is empty

leventeliu reviewed (approved):
LGTM.

ggicci committed:
Remove cqld client mode code

ggicci committed:
Remove Signee and Signature fields of models.Transaction

ggicci committed:
Don't use GConf in api package

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Fix RPC client unmarshal errror on Block.Timestamp

ggicci committed:
Fix bp_getTransactionList bug

ggicci committed:
Upgrade jsonrpc.Service's bootstrap workflow

1. add StopServersAndWait to Service instance;
2. export api.models.OpenSQLiteDBAsGorp function for convenient use;
3. update cqld's bootstrap.

ggicci committed:
Add test cases for jsonrpc service

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

leventeliu committed:
Merge branch 'feature/txInfo' into staging

leventeliu committed:
Add staging image tag

leventeliu committed:
Resolve FIXME

leventeliu committed:
Merge remote-tracking branch 'origin/feature/jsonrpc' into staging

ggicci committed:
Query transaction confirmed state

ggicci committed:
Register API node to BP DHT network

ggicci committed:
Merge branch 'staging' into fix/apiNodeDHTRegistry

ggicci committed:
Merge pull request #206 from CovenantSQL/fix/apiNodeDHTRegistry

Fix/api node dht registry

ggicci committed:
Fix API node register DHT failed

ggicci committed:
Suppress unused services in API node

ggicci committed:
Retry to connect to wsapi in TestJSONRPCService

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Fix API node bootstrap issue

1. use role Client to bootstrap the API node;
2. update BPGenesisInfo.Timestamp to a newer timestamp.

ggicci committed:
Add covenantsql_api_0 service in docker-compose.yml

leventeliu committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Remove unused HTTP stream transport

ggicci committed:
Refactor bp_getBlockList (pagination)

ggicci committed:
Add new API: bp_getTransactionListOfBlock

1. refactor bp_getTransactionList to support pagination;
2. add new API: bp_getTransactionListOfBlock.

ggicci committed:
Update integration test cases

ggicci committed:
Add missing utils file

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Refactor jsonrpc as a subpackage of rpc

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Fix missing call to api.StopService

ggicci committed:
Remove gotest -v option for gitlab CI

ggicci committed:
Fix data race on api.server instance

ggicci committed:
Add -coverpkg=all option to go test

ggicci committed:
Tweak -coverpkg option for gotest

ggicci committed:
Fix go test build error, split tests

ggicci committed:
Fix msgpack occasionally decode error

ggicci committed:
Upgrade gitlab CI config, use alltest.sh

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

ggicci committed:
Upgrade .gitlab-ci.yml

ggicci committed:
Format gocode

auxten commented on cmd/cqld/bootstrap.go:
```diff
@@ -77,6 +79,13 @@ func runNode(nodeID proto.NodeID, listenAddr string) (err error) {
 		return
 	}
 
+	if mode == "api" {
```
make "api" a const

auxten commented on cmd/cqld/main.go:
```diff
@@ -144,3 +153,51 @@ func main() {
 	log.Info("server stopped")
 }
+
+func registerNodeToBP(timeout time.Duration) (err error) {
```
why you copy this function from "cql-minerd"
if it's common usage why not to put it into some package

auxten commented on cmd/cqld/main.go:
```diff
@@ -76,6 +84,7 @@ func init() {
 	flag.StringVar(&metricWeb, "metric-web", "", "Address and port to get internal metrics")
 
+	flag.StringVar(&wsapiAddr, "wsapi", "", "Address of the websocket JSON-RPC API")
 	flag.StringVar(&mode, "mode", "normal", "Run mode, e.g. normal, api")
```
maybe just one flag "wsapi" is enough

ggicci committed:
Remove mode flag and other redundant code

1. remove mode flag from cqld command;
2. type define bp.RunMode, don't use string representation;
3. move registerNodeToBP to rpcutils.go;
4. remove redundant calls to registerNodeToBP.

ggicci committed:
Merge branch 'develop' into feature/jsonrpc

auxten reviewed (approved):
LGTM
